homeassistant/components/zamg/weather.py

```diff
@@ -52,93 +39,101 @@
-        _LOGGER.error(
-            "Configured ZAMG %s (%s) is not a known station",
-            CONF_STATION_ID,
-            station_id,
-        )
-        return
 
-    probe = ZamgData(station_id=station_id)
-    try:
-        probe.update()
-    except (ValueError, TypeError) as err:
-        _LOGGER.error("Received error from ZAMG: %s", err)
-        return
 
-    add_entities([ZamgWeather(probe, name)], True)
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the ZAMG weather platform."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [ZamgWeather(coordinator, entry.title, entry.data[CONF_STATION_ID])]
+    )
 
 
-class ZamgWeather(WeatherEntity):
+class ZamgWeather(CoordinatorEntity, WeatherEntity):
     """Representation of a weather condition."""
 
-    _attr_native_pressure_unit = PRESSURE_HPA
-    _attr_native_temperature_unit = TEMP_CELSIUS
-    _attr_native_wind_speed_unit = SPEED_KILOMETERS_PER_HOUR
-
-    def __init__(self, zamg_data, stationname=None):
+    def __init__(
+        self, coordinator: ZamgDataUpdateCoordinator, name, station_id
+    ) -> None:
         """Initialise the platform with a data instance and station name."""
-        self.zamg_data = zamg_data
-        self.stationname = stationname
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return (
-            self.stationname
-            or f"ZAMG {self.zamg_data.data.get('Name') or '(unknown station)'}"
+        super().__init__(coordinator)
+        self._attr_unique_id = f"{name}_{station_id}"
+        self._attr_name = f"ZAMG {name}"
+        self.station_id = f"{station_id}"
+        self._attr_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, station_id)},
+            manufacturer=ATTRIBUTION,
+            configuration_url=MANUFACTURER_URL,
+            name=coordinator.name,
         )
+        # set units of ZAMG API
+        self._attr_native_temperature_unit = TEMP_CELSIUS
+        self._attr_native_pressure_unit = PRESSURE_HPA
+        self._attr_native_wind_speed_unit = SPEED_METERS_PER_SECOND
+        self._attr_native_precipitation_unit = LENGTH_MILLIMETERS
 
     @property
-    def condition(self):
+    def condition(self) -> str | None:
         """Return the current condition."""
         return None
 
     @property
-    def attribution(self):
+    def attribution(self) -> str | None:
         """Return the attribution."""
         return ATTRIBUTION
 
     @property
-    def native_temperature(self):
+    def native_temperature(self) -> float | None:
         """Return the platform temperature."""
-        return self.zamg_data.get_data(ATTR_WEATHER_TEMPERATURE)
+        try:
+            return float(self.coordinator.data[self.station_id].get("TL")["data"])
```

Comment:
Why do we use `dict.get` and not `dict[key]` and catch `KeyError` instead of `TypeError`?

I'd consider restructuring the data in the coordinator instead during the coordinator update so we don't need to have all these guards in the platforms, and make it easier to look up a value from the `coordinator.data`.